src/main/java/org/folio/circulation/rules/cache/CirculationRulesCache.java

```diff
@@ -1,208 +1,123 @@
-      tenantId, isCurrent, rules.reloadTimestamp, currentTimestamp);
-    return isCurrent;
-  }
-
-  /**
-   * Reload is needed if the last reload is TRIGGER_AGE_IN_MILLISECONDS old
-   * and a reload hasn't been initiated yet.
-   * @param rules - rules to reload
-   * @return whether reload is needed
-   */
-  private boolean reloadNeeded(String tenantId, Rules rules) {
-    if (rules.reloadInitiated) {
-      log.info("Rules reload is already initiated for tenant {}", tenantId);
-      return false;
-    }
-
-    long currentTimestamp = System.currentTimeMillis();
-    boolean reloadNeeded = rules.reloadTimestamp + TRIGGER_AGE_IN_MILLISECONDS < currentTimestamp;
-    log.info("Rules reload is needed for tenant {}: {}. " +
-        "Reload timestamp is {}, current timestamp is {}",
-      tenantId, reloadNeeded, rules.reloadTimestamp, currentTimestamp);
-    return reloadNeeded;
-  }
+  private boolean rulesExist(String tenantId) {
+    Rules rules = rulesMap.get(tenantId);
 
-  private boolean rebuildNeeded(String tenantId, Rules rules) {
-    long currentTimestamp = System.currentTimeMillis();
-    boolean rebuildNeeded = rules.rebuildTimestamp + DROOLS_OBJECT_LIFETIME_IN_MILLISECONDS <
-      currentTimestamp;
-    log.info("Drools object rebuild is needed for tenant {}: {}. " +
-        "Rebuild timestamp is {}, current timestamp is {}",
-      tenantId, rebuildNeeded, rules.rebuildTimestamp, currentTimestamp);
-    return rebuildNeeded;
+    return rules != null && !rules.getRulesAsText().isEmpty();
   }
 
-  private CompletableFuture<Result<Rules>> reloadRules(String tenantId, Rules rules,
+  public CompletableFuture<Result<Rules>> reloadRules(String tenantId,
     CollectionResourceClient circulationRulesClient) {
-
     log.info("Reloading rules for tenant {}", tenantId);
 
     return circulationRulesClient.get()
-      .thenCompose(r -> r.after(response -> {
-        log.info("Fetched rules for tenant {}", tenantId);
-        JsonObject circulationRules = new JsonObject(response.getBody());
+      .thenApply(r -> r.map(response -> getRulesAsText(response, tenantId)))
+      .thenCompose(r -> r.after(rulesAsText -> reloadRules(tenantId, rulesAsText)));
+  }
 
-        rules.reloadTimestamp = System.currentTimeMillis();
-        rules.reloadInitiated = false;
+  private static String getRulesAsText(Response response, String tenantId) {
+    log.info("Fetched rules for tenant {}", tenantId);
 
-        if (log.isInfoEnabled()) {
-          log.info("circulationRules = {}", circulationRules.encodePrettily());
-        }
+    final var circulationRules = new JsonObject(response.getBody());
+    log.info("circulationRules = {}", circulationRules.encodePrettily());
+
+    return circulationRules.getString("rulesAsText");
+  }
 
-        String rulesAsText = circulationRules.getString("rulesAsText");
+  public CompletableFuture<Result<Rules>> reloadRules(String tenantId, String rulesAsText) {
```

Comment:
There is nothing asynchronous about this method, so no need to wrap the result into a Future

Reply:
I'm confused-this method is called as part of a promise chain that requires a future. I get that this isn't doing anything particularly asynchronous, but the chain expects a completed future.

Reply:
The chain can be changed to expect a `Result` instead of `CompletableFuture<Result>`:

`.thenCompose(r -> r.after(rulesAsText -> reloadRules(tenantId, rulesAsText)))`
becomes
`.thenApply(r -> r.next(rulesAsText -> reloadRules(tenantId, rulesAsText)))`